src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,1304 @@
+        void add_allocation(uint32_t result_type_id, uint32_t result_id, uint32_t storage_class, uint32_t initializer = 0);
+
+        std::map<Type, uint32_t> type_map;
+        std::map<std::pair<Type, uint32_t>, uint32_t> pointer_type_map;
+        std::map<Type, uint32_t> pair_type_map;
+        std::map<std::string, uint32_t> constant_map;
+
+        void add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                             std::initializer_list<uint32_t> words);
+        void add_instruction(uint32_t opcode, std::initializer_list<uint32_t> words);
+        void add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                             std::vector<uint32_t> words);
+        void add_instruction(uint32_t opcode, std::vector<uint32_t> words);
+        uint32_t map_type(const Type &type);
+        uint32_t map_pointer_type(const Type &type, uint32_t storage_class);
+        uint32_t map_type_to_pair(const Type &t);
+        uint32_t emit_constant(const Type &t, const void *data);
+        void scalarize(const Expr &e);
+
+        // The scope contains both the symbol and its storage class
+        Scope<std::pair<uint32_t, uint32_t>> symbol_table;
+
+        // The workgroup size.  Must be the same for all kernels.
+        uint32_t workgroup_size[3];
+
+        struct PhiNodeInputs {
+            uint32_t ids[4];
+        };
+        // Returns Phi node inputs.
+        template<typename StmtOrExpr>
+        PhiNodeInputs emit_if_then_else(const Expr &condition, StmtOrExpr then_case, StmtOrExpr else_case);
+    } emitter;
+
+    std::string current_kernel_name;
+};
+
+// --
+
+void CodeGen_Vulkan_Dev::SPIRVEmitter::add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                                                       std::initializer_list<uint32_t> words) {
+    region.push_back(((1 + words.size()) << 16) | opcode);
+    region.insert(region.end(), words.begin(), words.end());
+}
+
+void CodeGen_Vulkan_Dev::SPIRVEmitter::add_instruction(uint32_t opcode, std::initializer_list<uint32_t> words) {
+    spir_v_kernels.push_back(((1 + words.size()) << 16) | opcode);
+    spir_v_kernels.insert(spir_v_kernels.end(), words.begin(), words.end());
+}
+void CodeGen_Vulkan_Dev::SPIRVEmitter::add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                                                       std::vector<uint32_t> words) {
+    region.push_back(((1 + words.size()) << 16) | opcode);
+    region.insert(region.end(), words.begin(), words.end());
+}
+
+void CodeGen_Vulkan_Dev::SPIRVEmitter::add_instruction(uint32_t opcode, std::vector<uint32_t> words) {
+    spir_v_kernels.push_back(((1 + words.size()) << 16) | opcode);
+    spir_v_kernels.insert(spir_v_kernels.end(), words.begin(), words.end());
+}
+
+uint32_t CodeGen_Vulkan_Dev::SPIRVEmitter::emit_constant(const Type &t, const void *data) {
+    // TODO: this needs to emit OpConstantComposite for constants with lane > 1
```

Comment:
(Here and any/all remaining TODO comments:) Does this need a tracking issue? (If already planned to be fixed in subsequent phase, no; if open-ended someday-fix, then probably yes)